Visionfive2: Build bootable SD card image

a-firago opened the pull request:
This PR provides changes to create a bootable Visionfive2 SD card image, similar to the one created in the original Starfive SDK. 
Changes include:
* Add wks file with the partition layout taken from [here](https://github.com/starfive-tech/VisionFive2/blob/JH7110_VisionFive2_devel/conf/genimage-vf2.cfg)
* Add jh7710-spl-tool which is used to generate u-boot-spl.bin.normal.out expected on the SD card by the BootROM
* Deploy visionfive2_fw_payload.img, which is a FIT image containing fw_payload.bin expected by SPL started from SD card
* Update uEnv.txt, so that u-boot-starfive boots fitImage from the partition 3 and rootfs is expected on partition 4 of the SD card 

Current state - SD card flashed with the core-image-minimal-visionfive2.wic using 'dd' or 'bmaptool' boots fine with few issues:
* SPL built with bitbake, reports wrong DRAM size [here](https://github.com/starfive-tech/u-boot/blob/JH7110_VisionFive2_devel/arch/riscv/cpu/jh7110/dram.c#L14). Simply putting a debug message printing in the dram_init() function resolves this issue. Seems like this has something to do with compiler/flags. Investigating.
* During the boot, Linux kernel resets around 4 seconds timestamp. The same behavior is observed with the Starfive SDK [sdcard.img](https://github.com/starfive-tech/VisionFive2/releases/download/VF2_v2.10.4/sdcard.img). Investigating.

a-firago commented:
@cordlandwehr which boot source did you use with your PR https://github.com/riscv/meta-riscv/pull/382? Have you seen issues with SPL and/or kernel resets as described here?

kraj commented on recipes-bsp/opensbi/opensbi_%.bbappend:
```diff
@@ -1,2 +1,8 @@
 SRCREV:ae350-ax45mp = "22f38ee6c658a660083aa45c4ec6c72f66a17260"
+
+BRANCH:visionfive2 = "JH7110_VisionFive2_devel"
+SRC_URI:visionfive2 = "git://github.com/starfive-tech/opensbi.git;protocol=https;branch=${BRANCH}"
```
how different is this fork from upstream? is there a plan to upstream the changes can you comment on that ?

kraj commented on recipes-bsp/u-boot/u-boot-starfive_v2021.10.bb:
```diff
@@ -0,0 +1,42 @@
+    install -m 644 ${WORKDIR}/uEnv-visionfive2.txt ${DEPLOYDIR}/uEnv.txt
+}
+
+COMPATIBLE_MACHINE = "(visionfive2)"
```
does it support `visionfive1` as well? if so can we add that here as well, so we can have both machines share same u-boot.

a-firago replied:
JH7110_VisionFive2_devel branch of [github.com/starfive-tech/u-boot](https://github.com/starfive-tech/u-boot) moved to U-Boot v2021.10. Seems like this particular branch/tag is missing `visionfive1` support. I decided to create a new recipe because all branches supporting `visionfive1` are based on v2021.04.

kraj replied:
hmmm, this is not good, Can you check with upstream if there is a plan to maintain single branch for all visionfive variants ?

a-firago replied:
Asked here https://github.com/starfive-tech/u-boot/issues/43.

JH7110/Visionfive2 support in mainline U-Boot is currently under review (see https://rvspace.org/en/project/JH7110_Upstream_Plan and https://patchwork.ozlabs.org/project/uboot/list/?series=346514), so hopefully we can use upstream recipe soon.

Meanwhile, current u-boot-starfive_v2021.04 in meta-riscv is kind of broken:
* BRANCH for Visionfive is "JH7100_VisionFive_OH_dev" which is based on v2022.04
* BRANCH for Visionfive2  is "JH7110_VisionFive2_devel" which is based on v2021.10
* BRANCH for beaglev-starlight-jh7100 is "Fedora_JH7100_2021.04" which is based on  v2021.04

@kraj I guess the optimal solution now would be to rename "u-boot-starfive_v2021.04" to "u-boot-starfive" and set branches/SRCREV for each board.

kraj commented on recipes-bsp/jh7110-spl-tool/jh7110-spl-tool_git.bb:
```diff
@@ -0,0 +1,25 @@
+DESCRIPTION = "spl_tool is a jh7110 signature tool used to generate spl header information and generate u-boot-spl.bin.normal.out."
+LICENSE = "GPLv2"
```
Mark it `GPL-2.0-or-later`

a-firago replied:
ok

kraj commented on recipes-bsp/jh7110-spl-tool/jh7110-spl-tool_git.bb:
```diff
@@ -0,0 +1,25 @@
+do_install () {
+    cd ${S}
+    install -d ${D}${bindir}
+    install -m 0755 spl_tool ${D}${bindir}/spl_tool
```
whole function could be a single line
```
install -Dm 0755 ${S}/spl_tool ${D}${bindir}/spl_tool
```

a-firago replied:
ok

kraj commented on recipes-bsp/jh7110-spl-tool/jh7110-spl-tool_git.bb:
```diff
@@ -0,0 +1,25 @@
+DESCRIPTION = "spl_tool is a jh7110 signature tool used to generate spl header information and generate u-boot-spl.bin.normal.out."
+LICENSE = "GPLv2"
+LIC_FILES_CHKSUM = "file://LICENSE;md5=e6dc25dc2418b8831c906d43809d8336"
+SECTION = "BSP"
```
Make this to match u-boot
`SECTION = "bootloaders"`

a-firago replied:
ok

kraj commented on recipes-bsp/jh7110-spl-tool/jh7110-spl-tool_git.bb:
```diff
@@ -0,0 +1,25 @@
+    install -m 0755 spl_tool ${D}${bindir}/spl_tool
+}
+
+BBCLASSEXTEND = "native"
```
this seems redundant. You are already inheriting native on line 9

a-firago replied:
That's correct. But without `BBCLASSEXTEND = "native"` here, `DEPENDS:append:visionfive2 = " jh7110-spl-tool-native"` in U-Boot didn't work. I will re-check again

kraj replied:
this is not addressed in v2 ?

a-firago replied:
I've removed "inherit native" from line 9. Only BBCLASSEXTEND us kept.

cordlandwehr commented:
> @cordlandwehr which boot source did you use with your PR #382? Have you seen issues with SPL and/or kernel resets as described here?

I did all my tests by booting via the onboard SPL and than loading Kernel/rootfs from sd card. Right now I cannot recall any Kernel reset problems, but I am currently too many timezones away from my board to double check...

a-firago commented:
Updated PR:
* Implemented requested changes for spl_tool
* Used upstream opensbi - works the same.

Still in progress:
* proper fix for "Unsupport size" message in SPL
* Linux kernel resets around 4s after boot

a-firago commented:
* Added fix for SPL DRAM size reading from EEPROM;
* Kernel reset was an issue with a faulty USB TypeC cable on my setup. After cable replacement and SPL fix, SD card flashed with core-image-minimal.wic boots fine

@kraj please review.

kraj commented on recipes-bsp/u-boot/u-boot-starfive/fix_eeprom_read_issue.patch:
```diff
@@ -0,0 +1,20 @@
+diff --git a/board/starfive/visionfive2/visionfive2-i2c-eeprom.c b/board/starfive/visionfive2/visionfive2-i2c-eeprom.c
```
Can you add a patch header with some description of this patch and upstream status. For reference you can use this template

https://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines

kraj reviewed (changes requested):
please rebase the pr on top of latest master and push again after you have addressed the feedback. Thanks

a-firago commented:
Add v3:
* Updated U-Boot to tag v2.11.5
* Do not use 0001-riscv-fix-build-with-binutils-2.38.patch, it was merged upstream in v2021.10
* Remove fix EEPROM issue, it doesn't happen on v2.11.5

@kraj please review
